app/models/components/course/experience_points_records_ability_component.rb

```diff
@@ -0,0 +1,49 @@
+module Course::ExperiencePointsRecordsAbilityComponent
+  include AbilityHost::Component
+
+  def define_permissions
+    # cancancan is unable to merge abilities if administrators are not excluded
+    if user && !user.administrator?
+      allow_staff_manage_experience_points_records
+      allow_staff_and_points_owner_read_experience_points_records
+    end
+
+    super
+  end
+
+  private
+
+  def allow_staff_and_points_owner_read_experience_points_records
+    staff_roles = CourseUser::STAFF_ROLES.to_a.map { |role| CourseUser.roles[role] }
+    accessible_records  =
```

Comment:
...I don't understand what this query does. Break it up to scopes and use where only where necessary.

Are you sure that the hash does not work?